### ISSUE TITLE
Pose optimizing algorithm using image based loss doesn't change pose on each iteration

### ISSUE THREAD

**Issue by neanea04** [facebookresearch/theseus#595]
## ❓ Questions and Help
Description: I am trying to use the Theseus package to optimize the pose of an object from different 3D locations, so that the estimated pose from one camera observed from another camera is similar to the estimated pose from that camera. I am using a mean squared error (MSE) loss on the rendering of poses to optimize the pose.

However, I am experiencing an issue where the optimizing variables (v1, v2, v3, and v5) do not change on each iteration of the optimization. The initial pose given to the optimizer stays the same value for all the steps, and subsequently the LM optimization cannot find the optimized pose by minimizing the MSE error between two pictures.

I have tried running the optimization with different parameter values, but the optimizing variables still do not change. There are no error messages displayed during the optimization.

Here is a code snippet of the relevant functions:

  ```
  def detach_numpy(self,tensor):
        tensor = tensor.detach().cpu()
        if torch._C._functorch.is_gradtrackingtensor(tensor):
            
            tensor = torch._C._functorch.get_unwrapped(tensor)
            tensor = torch._C._functorch.get_unwrapped(tensor)
            return np.array(tensor.storage().tolist()).reshape(tensor.shape)
        return tensor.numpy()
    def cross_error(self,optim_vars, aux_vars):
        trans_list = []
        for t in optim_vars:
            xyz = self.detach_numpy(t.copy().to_x_y_z_quaternion())
            
            xyz = xyz.reshape((1,7))
            xyz = xyz[0]
            trans_list.append(xyz)


        inv_v1 = self.invert_transformation(trans_list[0])
        inv_v2 = self.invert_transformation(trans_list[1])

        v1 =trans_list[0]
        v2 = trans_list[1]
        v3 = trans_list[2]
        v5 = trans_list[3]

        
        #first contraint
        ren_list = [inv_v2,v1,v5]
        img1 = self.render_from_list(ren_list).clone().detach()

        ren_list = [v3]
        img2 = self.render_from_list(ren_list).clone().detach()

        error1 = self.mse_loss(img1, img2)



        #second contraint
        ren_list = [inv_v1,v2,v3]
        img1 = self.render_from_list(ren_list).clone().detach()

        ren_list = [v5]
        img2 = self.render_from_list(ren_list).clone().detach()

        error2 = self.mse_loss(img1, img2)

        error = error1 + error2
        error = error.unsqueeze(0).unsqueeze(0)

        error = np.random.rand(3,2)[0][0]*error.to("cpu")
        print(error)
        return error



    def quat_conv_conversion(self,quat,mode):
        if mode == "last2first":
            x,y,z,w = quat
            return np.array([w,x,y,z])
        if mode == "first2last":
            w,x,y,z = quat
            return np.array([x,y,z,w])
    
    def make_optim_var(self,node1,node2, name):
        var = self.calculate_relative_transformation(node1, node2)
        quat = (Rotation.from_matrix(var[:3,:3])).as_quat()
        quat = self.quat_conv_conversion(quat,"last2first")
        t = var[:3,3]
        var = np.array(list(t) + list(quat))
        var = torch.from_numpy(var)
        var = th.SE3(x_y_z_quaternion = var, name= name)

        return var

    def optimize(self):
        v1 = self.make_optim_var("world", "frame1","v1") 
        v2 = self.make_optim_var("world", "frame2","v2")
        v3 = self.make_optim_var("frame2", "obj1","v3")
        v5 = self.make_optim_var("frame1", "obj1","v5")


        

        optim_var = [v1,v2,v3,v5]     
        aux_var = []
        

        objective = th.Objective(torch.float64)
        cost_function = th.AutoDiffCostFunction(
            optim_vars = optim_var, err_fn= cast(ErrFnType,self.cross_error), dim =  1,aux_vars = aux_var,
            cost_weight=th.ScaleCostWeight(torch.tensor(1000, dtype=torch.float64)),autograd_mode="vmap")
        

        objective.add(cost_function)

        optimizer = th.LevenbergMarquardt(
        objective,
        max_iterations=10,
        step_size=1,
        linearization_cls=th.SparseLinearization,
        linear_solver_cls=th.CholmodSparseSolver,
        vectorize=False)

        
        inputs = {var.name: var for var in optim_var}

        optimizer.objective.update(inputs)

        start_event = torch.cuda.Event(enable_timing=True)
        end_event = torch.cuda.Event(enable_timing=True)

        start_event.record()
        torch.cuda.reset_peak_memory_stats()
        optimizer.optimize(verbose=True)
        end_event.record()

        torch.cuda.synchronize()
        forward_time = start_event.elapsed_time(end_event)
        forward_mem = torch.cuda.max_memory_allocated() / 1048576
```

Expected behavior:

The optimizing variables (v1, v2, v3, and v5) should change on each iteration of the optimization.
The LM optimization should converge to an optimized pose that minimizes the MSE error between two pictures.
Current behavior:

The optimizing variables do not change on each iteration of the optimization.
The LM optimization does not converge to an optimized pose.

Environment:

Operating system: Ubuntu 18.04
I would appreciate any help in resolving this issue. Thank you.

**Comment by luisenp** [comment]
Hi @neanea04, thanks for reporting. It's difficult to diagnose just by looking at code snippets. Have you checked if the jacobians computed by `AutodiffCostFunction` are non-zero? You can explore this by putting a breakpoint in [this](https://github.com/facebookresearch/theseus/blob/main/theseus/core/cost_function.py#L393) line.

You could also use
 
```python
optimizer.objective.update(inputs)
jacobians, error = cost_function.jacobians()
```

Probably unrelated, but the recommended way to use our optimizers is via a `TheseusLayer`, as explained in our [intro](https://github.com/facebookresearch/theseus/blob/main/tutorials/00_introduction.ipynb) tutorial. 

Also, you can reduce your profiling boilerplate with our [Timer](https://github.com/facebookresearch/theseus/blob/main/theseus/utils/utils.py#L203) class (see [example](https://github.com/facebookresearch/theseus/blob/main/examples/pose_graph/pose_graph_synthetic.py#L212)).

**Comment by neanea04** [comment]
The Jacobian seems to be zero vectors. Sorry I am new to optimization frameworks. As a simple explanation of my approach, I am just passing camera to object pose for few different cameras to the cost function. Using these poses inside the cost function I use my render function to render these poses. MSE loss between rendered images is used to optimize the camera pose. Is the jacobian zero because I am using a rendering function inside the cost function? The error output from the cost function seems to be alright.

**Comment by luisenp** [comment]
It looks like you are using a non-differentiable rendering function. I see you are doing some `numpy` conversions from the optimization variables, which would definite break the autodiff graph. You'll need a differentiable rendering function so that torch can compute the gradients automatically for you. 

@mhmukadam @fantaosha are you familiar with any libraries that can be used here for computing the error with differentiable rendering? Maybe `pytorch3d`?

**Comment by mhmukadam** [comment]
Good catch @luisenp! Yes PyTorch3D has differentiable rendering, maybe also nerfstudio.

**Comment by neanea04** [comment]
Thank you for your suggestions @luisenp @mhmukadam . I will try that.
